kyo-core/shared/src/main/scala/kyo/Channel.scala

```diff
@@ -180,7 +210,7 @@ object Channel:
         )(using initFrame: Frame, allow: AllowUnsafe): Unsafe[A] =
             new Unsafe[A]:
                 val queue = Queue.Unsafe.init[A](_capacity, access)
-                val takes = new MpmcUnboundedXaddArrayQueue[Promise.Unsafe[Closed, A]](8)
+                val takes = new MpmcUnboundedXaddArrayQueue[Promise.Unsafe[Closed, A] | (Int, Promise.Unsafe[Closed, Seq[A]])](8)
```

Comment:
`takes` element can now be either a request for a single element or for multiple elements as specified by first element of tuple

Reply:
I'm not sure a `drain` operation should block. Isn't just returning what's currently in the channel like in `drain` enough? The outer loop in `takeExactly` can then keep taking more items if the initial drain doesn't have enough elements.

Reply:
The problem is that if we loop over `drain`/`drainUpTo`, each iteration will return with an empty `Seq` when the channel's empty.  I figure we want to avoid unnecessary polling here.

Reply:
I had in mind an initial non-blocking drain and then a loop that uses take to block if needed to wait for more results

Reply:
So after initial drain, just take one at a time? Isn't it more efficient to take in chunks? Or is having tuples in the `takes` queue more overhead than is worth it?

Reply:
My initial concern was more semantic. The current API separates `take` as an operation that may park with `Async` from `drain` that returns synchronously given the `IO`-only restriction. This distinction is useful to reason about how to compose these operations efficiently. It might be more a question of method naming than semantics, though.

Regarding performance, I wouldn't be surprised if taking items individually has similar performance to a batched operation. There's the tuple allocation you mentioned, the additional branching in the channel flushing logic, and the memory barriers to modify the queue are needed for each item individually anyway. Perhaps `takeExactly` could be a loop that starts with an initial `drainUpTo` and then intercalates `take` and `drainUpTo` on each iteration?

Reply:
We should probably follow up with batched channel insertion. It could be a good performance improvement if paired with this change.

Reply:
Ah good idea. I'll do that.